File: python-setup/search_tools/wikipedia.py
import requests

def query(
  query: str,
):
  """
  Use this tool to search for information on a places from Wikipedia. Else don't use this tool.

  Args:
      query: the name of object or topic to find.

  Example: {"answer": "Southeast Asia is home to over 650 million people, representing hundreds of ethnic groups and languages. Major countries like Indonesia, the Philippines, and Vietnam each have distinct cultures and traditions."}
  """

  try:
    wiki_title = search_wiki_title(query)
    wiki_full_text = get_wiki_full_text(wiki_title)
  except Exception as e:
    return {"answer": "Sorry, I couldn't find any information on that topic."}

  return {"answer": wiki_full_text}

# Search for a relevant wikipedia article title
def search_wiki_title(query):
  url = f"https://en.wikipedia.org/w/api.php?action=query&list=search&srsearch={query}&srlimit=1&format=json"
  resp = requests.get(url)
  return resp.json()["query"]["search"][0]["title"]

# From previously found title, get the article text
def get_wiki_full_text(wiki_title):
  url = f"https://en.wikipedia.org/w/api.php?action=query&prop=extracts&titles={wiki_title}&explaintext=true&format=json"
  response = requests.get(url)
  data = response.json()
  page_id = next(iter(data["query"]["pages"]))
  plain_text = data["query"]["pages"][page_id]["extract"]
  return plain_text

File: node-serverless/server/gemini_functions/wikipedia.js
const { FunctionDeclarationSchemaType } = require('@google-cloud/vertexai');
const axios = require('axios');

const getWikipediaContentFunctionDeclaration = {
  name: 'get_wikipedia_content',
  description: 'Wikipedia is an online encyclopedia that provides information on a wide range of topics. This function will return the content of a Wikipedia article based on the subject provided.',
  parameters: {
    type: FunctionDeclarationSchemaType.OBJECT,
    properties: {
      subject: {
        type: FunctionDeclarationSchemaType.STRING,
        description: 'The name of object or topic to find.'
      },
    },
    required: ['subject'],
  },
};

async function getWikipediaContent(subject) {
  try {
    const subjectTitle = await searchWikiTitle(subject);
    if (subjectTitle) {
      const answer = await getWikiFullText(subjectTitle);
      return answer;
    }

    return null;
  } catch (e) {
    console.error(e);
    return null;
  }
}

// Search for a relevant wikipedia article title
async function searchWikiTitle(subject) {
  const response = await axios.get('https://en.wikipedia.org/w/api.php', {
    params: {
      action: 'query',
      list: 'search',
      srsearch: subject,
      srlimit: 1,
      format: 'json',
    }
  });

  if (response?.data?.query?.search &&
    response?.data?.query?.search.length > 0 &&
    response?.data?.query?.search[0]?.title) {

    return response.data.query.search[0].title
  }

  return null;
}

// From previously found title, get the article text
async function getWikiFullText(wikiTitle){
  const response = await axios.get('https://en.wikipedia.org/w/api.php', {
    params: {
      action: 'query',
      prop: 'extracts',
      titles: wikiTitle,
      explaintext: 'true',
      format: 'json',
    }
  });

  if (response?.data?.query?.pages) {
    const keys = Object.keys(response?.data?.query?.pages);
    const pages = response?.data?.query?.pages;

    if (keys.length > 0 && pages[keys[0]]?.extract) {
      return pages[keys[0]].extract;
    }
  }

  return null;
}

module.exports = {
  getWikipediaContentFunctionDeclaration,
  getWikipediaContent,
};
